blob/azureblob/azureblob.go

```diff
@@ -0,0 +1,586 @@
+}
+
+func (b *bucket) SignedURL(ctx context.Context, key string, opts *driver.SignedURLOptions) (string, error) {
+
+	key = strings.Replace(key, OSPathSeparator, b.defaultDelimiter, -1)
+
+	blockBlobURL := b.urls.containerURL.NewBlobURL(key)
+	srcBlobParts := azblob.NewBlobURLParts(blockBlobURL.URL())
+	var err error
+	srcBlobParts.SAS, err = azblob.BlobSASSignatureValues{
+		Protocol:      azblob.SASProtocolHTTPS,
+		ExpiryTime:    time.Now().UTC().Add(opts.Expiry),
+		ContainerName: b.name,
+		BlobName:      key,
+		Permissions:   azblob.BlobSASPermissions{Add: true, Create: true, Delete: true, Read: true, Write: true}.String(),
+	}.NewSASQueryParameters(b.credential)
+
+	if err != nil {
+		return "", makeAzureError(err, b.name, key)
+	}
+
+	srcBlobURLWithSAS := srcBlobParts.URL()
+	return srcBlobURLWithSAS.String(), nil
+}
+
+type writer struct {
+	ctx         context.Context
+	urls        *serviceUrls
+	key         string
+	contentType string
+	blockIDs    []string
+	mux         sync.Mutex
+	writerOpts  *driver.WriterOptions
+}
+
+// NewTypedWriter implements driver.NewTypedWriter.
+func (b *bucket) NewTypedWriter(ctx context.Context, key string, contentType string, opts *driver.WriterOptions) (driver.Writer, error) {
+
+	if key == "" {
+		return nil, fmt.Errorf("Invalid/Empty Key")
+	}
+
+	key = strings.Replace(key, OSPathSeparator, b.defaultDelimiter, -1)
+
+	containerURL := b.urls.serviceURL.NewContainerURL(b.name)
+	// try to create the azure container
+	// if it already exists, continue..
+	// if authorization failure, assume SASToken scope is limited to the Blob and continue..
+	// fail all other exceptions
+	_, err := containerURL.Create(ctx, nil, b.PublicAccessType)
+	if err != nil {
+		if serr, ok := err.(azblob.StorageError); ok {
+			switch serr.ServiceCode() {
+			case azblob.ServiceCodeContainerAlreadyExists:
+				// Can be thrown if container already exist, ignore and continue
+			case azblob.ServiceCodeAuthenticationFailed:
+				fallthrough
+			case "AuthorizationFailure":
+				// Can be thrown if SASToken Scope is restricted to Read & Write on Blobs
+				// Assume container exist, blob operations will succeed/fail based on SASToken scope
+			default:
+				return nil, azureError{bucket: b.name, key: key, msg: err.Error()}
+			}
+		}
+	}
+
+	key = strings.Replace(key, OSPathSeparator, b.defaultDelimiter, -1)
+	blockBlobURL := containerURL.NewBlockBlobURL(key)
+
+	var blockIDs []string
+	w := &writer{
+		ctx:         ctx,
+		key:         key,
+		contentType: contentType,
+		writerOpts:  opts,
+		urls: &serviceUrls{
+			serviceURL:   b.urls.serviceURL,
+			containerURL: &containerURL,
+			blockBlobURL: &blockBlobURL,
+		},
+		blockIDs: blockIDs,
+	}
+
+	return w, nil
+}
+
+// Write creates a stated block for incoming buffer (p)
+// Each call to Write will append to the blockId list for final commit in w.Close()
+func (w *writer) Write(p []byte) (int, error) {
+	chunks := split(p, azblob.BlockBlobMaxStageBlockBytes)
```

Comment:
Hmm, I don't think that chunking here will be effective, since the user is controlling how many bytes they are passing to each `Write`. They may make hundreds or thousands of calls to `Write` with a few bytes each; we don't want to do a separate request for each one.

Can you use [Upload](https://godoc.org/github.com/Azure/azure-storage-blob-go/azblob#BlockBlobURL.Upload) or [UploadStreamToBlockBlob](https://godoc.org/github.com/Azure/azure-storage-blob-go/azblob#UploadStreamToBlockBlob) instead? See `s3blob` for an example that may be relevant.

Reply:
I see your point. It was originally implemented with Pipes similar to S3’s, but I was persuaded to use Stage/Commit pattern. I will revert using IO Pipes. As an FYI, UploadStreamToBlob is chunking and using Stage/Commit however, it will do so with a full buffer when Close is called.

Reply:
You may have to include a special request for zero-length write. I think we did something similar for S3: https://github.com/google/go-cloud/blob/master/blob/s3blob/s3blob.go#L206.

Reply:
After a lot of testing, I have narrowed this failure to the recorder and this line of code. The original request has a type http.noBody which is converted to a nopCloser. This is somehow causing the Azure API to reject the request with the above response. Thoughts?

[from go-vcr](https://github.com/dnaeon/go-vcr/blob/cd28cbc9b4cb410879adb3670e993cc545f2a167/recorder/recorder.go#L101)
`if r.Body != nil {
        // Record the request body so we can add it to the cassette
        r.Body = ioutil.NopCloser(io.TeeReader(r.Body, reqBody))
    }
`

Reply:
Could be related to [this ](https://github.com/golang/go/issues/20257) issue

Reply:
I still see your code manually doing chunking. Why can't you just use [UploadStreamToBlockBlob](https://godoc.org/github.com/Azure/azure-storage-blob-go/azblob#UploadStreamToBlockBlob) and give it an `io.Reader` that's the other end of the pipe?

Or have you not pushed the updated code yet?

Does `Upload` work for a zero-length blob? You may need to special case 0-length if the API you're normally using (hopefully `UploadStreamToBlockBlob`) doesn't work for it.

Reply:
The updated code is using UploadStreamToBlockBlob. UploadStreamToBlockBlob will chunk if the content size is greater than BufferSize otherwise it performs a simple Upload. The issue I am facing is when Upload is called in Recording mode, the recorder changes the http request’s emptyBody to nopCloser causing the above MissingContentLengthHeader. The issue appears to be because of [this](https://github.com/golang/go/issues/20257)

The only way around this issue is to stage/commit empty blobs myself to avoid UploadStreamToBlockBlob in this scenario . Of course, this is only needed because of the recorder.

Reply:
Can you leave the Body `nil` in this scenario? It looks like the replay code you pointed to only adds a nopCloser if `Body != nil`.

Sorry about the troubles with go-vcr, that's frustrating. I have sent them patches before and they were accepted quickly, so an alternative here if you think there's a bug in the replay code is to fix it.

Reply:
Now that I'm able to run the tests locally, I've been looking into this. It does look like a bug in go-vcr; I sent a PR to fix it (https://github.com/dnaeon/go-vcr/pull/39).

Interestingly, as part of this I noticed #1019: your `touch` function, modeled after `s3blob`'s, is never being called, because the concrete type always inserts a 0-length write when `Close` is called with no `Write`s. This may have been making your debugging harder.